File: templates/websocket_template/src/core_specs/configuration/config_loader.py
"""
#############################################################################
### Configuration loader file
###
### @file config_loader.py
### @author Sebastian Russo
### @date 2025
#############################################################################

This module loads configuration data from a JSON file.
"""

# Native imports
import json
import sys


def read_data_from_config_json(file_path: str, exit_on_error: bool = True) -> dict:
    """
    Reads data from a JSON configuration file.

    Parameters:
        file_path (str): Path to the JSON config file.
        exit_on_error (bool): Whether to exit on error or return None.

    Returns:
        dict: Parsed JSON configuration data.
    """
    try:
        with open(file_path, "r", encoding="utf-8") as file:
            config_data = json.load(file)
        return config_data
    except FileNotFoundError:
        print(f"ERROR: Config file not found: {file_path}")
        if exit_on_error:
            sys.exit(1)
        return None
    except json.JSONDecodeError:
        print(f"ERROR: Failed to parse JSON config file: {file_path}")
        if exit_on_error:
            sys.exit(1)
        return None


CONFIG_FILE_PATH = "src/core_specs/configuration/config_file.json"
config_loader = read_data_from_config_json(CONFIG_FILE_PATH, exit_on_error=True)

File: templates/java_rest_api_template/src/main/java/com/template/restapi/core_specs/configuration/ConfigLoader.java
/**
 * #############################################################################
 * ### Configuration loader file
 * ###
 * ### @file ConfigLoader.java
 * ### @author Sebastian Russo
 * ### @date 2026
 * #############################################################################
 *
 * Loads configuration data from config_file.json (on the classpath) once, at
 * class-initialization time, and exposes it as a parsed JsonNode tree for the
 * rest of the application.
 *
 *   Java:    ConfigLoader.endpoint("example_endpoint_1").get("endpoint_route").asText()
 *
 * Note: CoreSpecsInitializer additionally bridges this same JSON into the
 * Spring Environment so the values can be referenced via ${config....}
 * placeholders inside controller annotations.
 */
package com.template.restapi.core_specs.configuration;

import java.io.InputStream;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

public final class ConfigLoader {

    /** Classpath location of the config file (under src/main/resources). */
    private static final String CONFIG_PATH = "/core_specs/configuration/config_file.json";

    private static final JsonNode CONFIG = load();

    private ConfigLoader() { }

    /**
     * Reads and parses config_file.json. Mirrors read_data_from_config_json():
     * a failure here is fatal because nothing can run without configuration.
     */
    private static JsonNode load() {
        try (InputStream in = ConfigLoader.class.getResourceAsStream(CONFIG_PATH)) {
            if (in == null) {
                System.err.println("ERROR: Config file not found on classpath: " + CONFIG_PATH);
                System.exit(1);
            }
            return new ObjectMapper().readTree(in);
        } catch (Exception ex) {
            System.err.println("ERROR: Failed to parse JSON config file '" + CONFIG_PATH + "': " + ex.getMessage());
            System.exit(1);
            return null; // unreachable; keeps the compiler happy
        }
    }

    /** The entire parsed configuration tree (equivalent to the config_loader dict). */
    public static JsonNode get() {
        return CONFIG;
    }

    /** Convenience accessor for the "network" section. */
    public static JsonNode network() {
        return CONFIG.path("network");
    }

    /** Convenience accessor for the "logging" section. */
    public static JsonNode logging() {
        return CONFIG.path("logging");
    }

    /** Convenience accessor for the "email_validation" section. */
    public static JsonNode emailValidation() {
        return CONFIG.path("email_validation");
    }

    /** Convenience accessor for a single endpoint entry under "endpoints". */
    public static JsonNode endpoint(String key) {
        return CONFIG.path("endpoints").path(key);
    }
}
